### ISSUE TITLE
[Feature] Restrict traffic to IMDS endpoints

### ISSUE THREAD

**Issue by CocoWang-wql** [Azure/AKS#4037]
Work on a solution to block the traffic from most pods to AKS node IMDS restriction. And users can set some allowlist rules.

**Comment by esierra-stratio** [comment]
Is it possible to achieve this using Calico network policies? I have attempted to block all egress traffic to the IMDS (169.254.169.254) or the NMI namespace, but haven't succeeded yet.

**Comment by CocoWang-wql** [comment]
We are working on a managed solution, with which, the users are released/managed to set network policies manually.

**Comment by CocoWang-wql** [comment]
We hope to make the feature public preview at the end of August!

**Comment by jkroepke** [comment]
@CocoWang-wql does the managed solution require a Network Policy engine like calico or cilium?

**Comment by charleswool** [comment]
Update, we found an issue during bugbash, the public preview date is postponed to end of Sep

**Comment by mdnix** [comment]
What's the current state of this?

**Comment by charleswool** [comment]
IMDS Restriction is now public preview

**Comment by esierra-stratio** [comment]
Once enabled, is it possible to make exceptions to allow certain services within the cluster to access that endpoint?

**Comment by jkroepke** [comment]
@esierra-stratio otherwise, disable the managed feature and use classical network policies for a more fine-granted approach.

**Comment by charleswool** [comment]
Yeah Terraform only support GA feature. With that being said, we have the ETA for IMDS restriction GA set as Sep CY25 at this moment

**Comment by gravufo** [comment]
I'm not sure why this feature was even implemented if it would not even work with Microsoft's own add-ons. The sheer length of the [limitations section](https://learn.microsoft.com/en-us/azure/aks/imds-restriction#limitations) in the documentation page should have been a red flag when developping this feature.
We will continue using our own custom network policies.

**Comment by charleswool** [comment]
@gravufo We are working to shorten that limitation list by migration these add-ons to support workload identity. In this way these pod can authenticate with Entra for being able to access Azure resources or service data plane endpoints and no longer require access to IMDS endpoint. The ETA for complete is September.

**Comment by gravufo** [comment]
@charleswool thank you for the feedback and timeline.
